### PR TITLE
SuccessFactors RunTime Implementation Part II

### PR DESCRIPTION
**Pull request by sgarg-CS**

### CODE REVIEW

File: src/main/java/io/cdap/plugin/successfactors/common/exception/proto/innererror/ErrorResolution.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+/**
+ * ErrorResolution class
+ */
+public class ErrorResolution {
+
+  private final String sapTransaction;
+  private final String sapNote;
+  private final String additionalSAPNote;
+
+  public ErrorResolution(String sapTransaction, String sapNote, String additionalSAPNote) {
```
Comment (vikasrathee-cs) [inline src/main/java/io/cdap/plugin/successfactors/common/exception/proto/innererror/ErrorResolution.java]: can we have more successfactors specific name?

File: src/main/java/io/cdap/plugin/successfactors/common/util/ExceptionParser.java [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+      String rawResponseString = "";
+      InputStream rawStream = responseContainer.getResponseStream();
+      if (rawStream != null) {
+        try (BufferedReader bufferedReader = new BufferedReader(
+          new InputStreamReader(rawStream, StandardCharsets.UTF_8))) {
+
+          rawResponseString = bufferedReader.lines().collect(Collectors.joining(" "));
+          // For handling broken hyperlinks to SAP docs in error messages returned by SAP SuccessFactors OData v2 API
+          if (rawResponseString.contains("Refer to https")) {
+            rawResponseString = rawResponseString.substring(0, rawResponseString.indexOf("Refer to https"));
```
Comment (vikasrathee-cs) [inline src/main/java/io/cdap/plugin/successfactors/common/util/ExceptionParser.java]: create a variable for "Refer to https"

File: src/main/java/io/cdap/plugin/successfactors/common/util/ExceptionParser.java [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+          }
+        } catch (IOException ioe) {
+          // no-ops
```
Comment (vikasrathee-cs) [inline src/main/java/io/cdap/plugin/successfactors/common/util/ExceptionParser.java]: Don't we need to throw this exception?

File: src/main/java/io/cdap/plugin/successfactors/common/util/ExceptionParser.java [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+      }
+
+      LOG.error("HTTP Code: {}", responseContainer.getHttpStatusCode());
+      LOG.error("Detailed error message: {} {}", msg, rawResponseString);
+
+      try {
+        error = GSON.fromJson(rawResponseString, SuccessFactorsError.class);
+      } catch (JsonSyntaxException | JsonIOException je) {
+        // html errors are only found in case of invalid SuccessFactors service namespace
+        if (rawResponseString.startsWith("<html>")) {
```
Comment (vikasrathee-cs) [inline src/main/java/io/cdap/plugin/successfactors/common/util/ExceptionParser.java]: create a variable

File: src/main/java/io/cdap/plugin/successfactors/source/config/SuccessFactorsPluginConfig.java [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+  /**
+   * Validates the mandatory parameters.
+   *
+   * @param failureCollector {@code FailureCollector}
+   */
+  private void validateMandatoryParameters(FailureCollector failureCollector) {
+
+    IdUtils.validateReferenceName(getReferenceName(), failureCollector);
+    if (SuccessFactorsUtil.isNullOrEmpty(getBaseURL()) && !containsMacro(BASE_URL)) {
+      String errMsg = ResourceConstants.ERR_MISSING_PARAM_PREFIX.getMsgForKey("SAP SuccessFactor Base URL");
```
Comment (vikasrathee-cs) [inline src/main/java/io/cdap/plugin/successfactors/source/config/SuccessFactorsPluginConfig.java]: use variable for SAP SuccessFactor Base URL

File: src/main/java/io/cdap/plugin/successfactors/source/config/SuccessFactorsPluginConfig.java [hunk 6]
```diff
@@ -0,0 +1,371 @@
...
+      failureCollector.addFailure(errMsg, COMMON_ACTION).withConfigProperty(BASE_URL);
+    }
+    if (SuccessFactorsUtil.isNotNullOrEmpty(getBaseURL()) && !containsMacro(BASE_URL)) {
+      if (HttpUrl.parse(getBaseURL()) == null) {
+        String errMsg = ResourceConstants.ERR_INVALID_BASE_URL.getMsgForKey("SAP SuccessFactor Base URL");
+        failureCollector.addFailure(errMsg, COMMON_ACTION).withConfigProperty(BASE_URL);
+      }
+    }
+    if (SuccessFactorsUtil.isNullOrEmpty(getEntityName()) && !containsMacro(ENTITY_NAME)) {
+      String errMsg = ResourceConstants.ERR_MISSING_PARAM_PREFIX.getMsgForKey("Entity Name");
```
Comment (vikasrathee-cs) [inline src/main/java/io/cdap/plugin/successfactors/source/config/SuccessFactorsPluginConfig.java]: Use variable and use the same in enum class also

File: src/main/java/io/cdap/plugin/successfactors/source/config/SuccessFactorsPluginConfig.java [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+
+  /**
+   * Validates the credentials parameters.
+   *
+   * @param failureCollector {@code FailureCollector}
+   */
+  private void validateBasicCredentials(FailureCollector failureCollector) {
+
+    if (SuccessFactorsUtil.isNullOrEmpty(getUsername()) && !containsMacro(UNAME)) {
+      String errMsg = ResourceConstants.ERR_MISSING_PARAM_PREFIX.getMsgForKey("SAP SuccessFactors Username");
```
Comment (vikasrathee-cs) [inline src/main/java/io/cdap/plugin/successfactors/source/config/SuccessFactorsPluginConfig.java]: use variable for all getMsgForKey parameters

File: src/main/java/io/cdap/plugin/successfactors/source/input/SuccessFactorsInputFormat.java [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+      .get(SUCCESSFACTORS_PLUGIN_PROPERTIES), SuccessFactorsPluginConfig.class);
+
+    Schema outputSchema = Schema.parseJson(taContext.getConfiguration().get(SuccessFactorsSource.OUTPUT_SCHEMA));
+
+    String encodedMetadataString = taContext.getConfiguration().get(ENCODED_ENTITY_METADATA_STRING);
+
+    SuccessFactorsTransporter transporter = new SuccessFactorsTransporter(pluginConfig.getUsername(),
+                                                                          pluginConfig.getPassword());
+
+    SuccessFactorsService successFactorsService = new SuccessFactorsService(pluginConfig, transporter);
```
Comment (vikasrathee-cs) [inline src/main/java/io/cdap/plugin/successfactors/source/input/SuccessFactorsInputFormat.java]: Write a separate method to get SuccessFactorsService and reuse it everywhere.

File: src/main/java/io/cdap/plugin/successfactors/source/input/SuccessFactorsPartitionBuilder.java [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+  public List<SuccessFactorsInputSplit> buildSplits(long availableRecordCount) {
+
+    List<SuccessFactorsInputSplit> list = new ArrayList<>();
+    long start = 1;
+    // setting up the optimal split size and count values
+    long batchSize = Math.min(availableRecordCount, MAX_ALLOWED_BATCH_SIZE);
+    long optimalLoadOnSplit = Math.min(availableRecordCount, MAX_RECORDS_IN_SPLIT);
+    long optimalSplitCount = availableRecordCount / optimalLoadOnSplit +
+      (availableRecordCount % optimalLoadOnSplit != 0 ? 1 : 0);
+    long leftoverLoadCount = availableRecordCount % optimalSplitCount;
```
Comment (vikasrathee-cs) [inline src/main/java/io/cdap/plugin/successfactors/source/input/SuccessFactorsPartitionBuilder.java]: Not able to understand this logic, need to discuss on this.

File: src/test/java/io/cdap/plugin/successfactors/source/config/SuccessFactorsPluginConfigTest.java [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+  private SuccessFactorsPluginConfig.Builder pluginConfigBuilder;
+
+  @Before
+  public void setUp() {
+    failureCollector = new MockFailureCollector();
+    pluginConfigBuilder = SuccessFactorsPluginConfig.builder()
+      .referenceName("unit-test-ref-name")
+      .baseURL("http://localhost")
+      .entityName("entity name")
+      .username("username")
```
Comment (vikasrathee-cs) [inline src/test/java/io/cdap/plugin/successfactors/source/config/SuccessFactorsPluginConfigTest.java]: use variables

File: src/main/java/io/cdap/plugin/successfactors/common/exception/proto/innererror/ErrorResolution.java (hunk 1 above)
Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/successfactors/common/exception/proto/innererror/ErrorResolution.java]: Fields have been renamed appropriately. 4dbd00c9806f47df7f78cf411834bc8c7a613f94

File: src/main/java/io/cdap/plugin/successfactors/common/util/ExceptionParser.java (hunk 2 above)
Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/successfactors/common/util/ExceptionParser.java]: Added `BROKEN_HYPERLINK` for "Refer to https" 4dbd00c9806f47df7f78cf411834bc8c7a613f94

File: src/main/java/io/cdap/plugin/successfactors/common/util/ExceptionParser.java (hunk 3 above)
Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/successfactors/common/util/ExceptionParser.java]: throwing SuccessFactorsServiceException 4dbd00c9806f47df7f78cf411834bc8c7a613f94

File: src/main/java/io/cdap/plugin/successfactors/common/util/ExceptionParser.java (hunk 4 above)
Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/successfactors/common/util/ExceptionParser.java]: Added `HTML_TAG` for` "<html>"`  4dbd00c9806f47df7f78cf411834bc8c7a613f94

File: src/main/java/io/cdap/plugin/successfactors/source/config/SuccessFactorsPluginConfig.java (hunk 5 above)
Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/successfactors/source/config/SuccessFactorsPluginConfig.java]: Added. 4dbd00c9806f47df7f78cf411834bc8c7a613f94

File: src/main/java/io/cdap/plugin/successfactors/source/config/SuccessFactorsPluginConfig.java (hunk 6 above)
Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/successfactors/source/config/SuccessFactorsPluginConfig.java]: Added. 4dbd00c9806f47df7f78cf411834bc8c7a613f94

File: src/main/java/io/cdap/plugin/successfactors/source/config/SuccessFactorsPluginConfig.java (hunk 7 above)
Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/successfactors/source/config/SuccessFactorsPluginConfig.java]: Added constants for all strings. 4dbd00c9806f47df7f78cf411834bc8c7a613f94

File: src/main/java/io/cdap/plugin/successfactors/source/input/SuccessFactorsPartitionBuilder.java (hunk 9 above)
Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/successfactors/source/input/SuccessFactorsPartitionBuilder.java]: Refactored the `buildSplits()` method as per our discussion. 4dbd00c9806f47df7f78cf411834bc8c7a613f94

File: src/test/java/io/cdap/plugin/successfactors/source/config/SuccessFactorsPluginConfigTest.java (hunk 10 above)
Comment (sgarg-CS) [reply src/test/java/io/cdap/plugin/successfactors/source/config/SuccessFactorsPluginConfigTest.java]: Added. 4dbd00c9806f47df7f78cf411834bc8c7a613f94

File: src/main/java/io/cdap/plugin/successfactors/source/input/SuccessFactorsInputFormat.java (hunk 8 above)
Comment (sgarg-CS) [reply src/main/java/io/cdap/plugin/successfactors/source/input/SuccessFactorsInputFormat.java]: Created a separate method : c60d7165923dc55eab2e5a3dcc247cd770075b47

**Pull request merged**